include/util/hex_float.h

```diff
@@ -213,12 +270,310 @@ class HexFloat {
   static const uint32_t fraction_right_shift =
       (sizeof(uint_type) * 8) - num_fraction_bits;
 
+  // The maximum representable normalized exponent.
+  static const int_type max_exponent =
+      (exponent_mask >> num_fraction_bits) - exponent_bias;
+  // The minimum representable exponent for normalized numbers.
+  static const int_type min_exponent = -exponent_bias;
+
+  // Returns the bits associated with the value.
+  uint_type getBits() const {
+    return spvutils::BitwiseCast<uint_type>(value_);
+  }
+
+  // Returns the bits associated with the value, without the leading sign bit.
+  uint_type getUnsignedBits() const {
+    return spvutils::BitwiseCast<uint_type>(value_) & ~sign_mask;
+  }
+
+  // Returns the bits associated with the exponent, shifted to start at the
+  // lsb of the type.
+  const uint_type getExponentBits() const {
+    return (getBits() & exponent_mask) >> num_fraction_bits;
+  }
+
+  // Returns the exponent in unbiased form. This is the exponent in the
+  // human-friendly form.
+  const int_type getUnbiasedExponent() const {
+    return (getExponentBits() - exponent_bias);
+  }
+
+  // Returns just the significand bits from the value.
+  const uint_type getSignificandBits() const {
+    return getBits() & fraction_encode_mask;
+  }
+
+  // If the number was normalized, returns the unbiased exponent.
+  // If the number was denormal, normalize the exponent first.
+  const int_type getUnbiasedNormalizedExponent() const {
+    uint_type non_sign_bits = getBits() & ~sign_mask;
+    if ((non_sign_bits & ~sign_mask) == 0) {  // special case if everything is 0
+      return 0;
+    }
+    int_type exp = getUnbiasedExponent();
+    if (exp == -exponent_bias &&
+        non_sign_bits != 0) {  // We are in denorm land.
+      uint_type significand_bits = getSignificandBits();
+      while ((significand_bits & (first_exponent_bit >> 1)) == 0) {
+        significand_bits <<= 1;
+        exp -= 1;
+      }
+      significand_bits &= fraction_encode_mask;
+    }
+    return exp;
+  }
+
+  // Returns the signficand after it has been normalized.
+  const uint_type getNormalizedSignificand() const {
+    int_type unbiased_exponent = getUnbiasedNormalizedExponent();
+    uint_type significand = getSignificandBits();
+    for (int_type i = unbiased_exponent;
+         i <= -static_cast<int_type>(exponent_bias); ++i) {
+      significand <<= 1;
+    }
+    significand &= fraction_encode_mask;
+    return significand;
+  }
+
+  // Returns true if this number represents a negative value.
+  bool isNegative() const { return (getBits() & sign_mask) != 0; }
+
+  // Note this assumes EVERY significand is normalized, and has an implicit
+  // leading one. This means that the only way that this method will set 0,
```

Comment:
Interesting.   So this method should not be used with raw bits extracted from a subnormal number, since only normal numbers have the implicit leading 1.  I'd add something to that effect after the "This means that".  E.g.
"Also, don't use this method with raw bits extracted from a subnormal number, since subnormals don't have leading implicit 1 in the significand."

 https://en.wikipedia.org/wiki/Half-precision_floating-point_format